beluga/include/beluga/motion/differential_drive_model.hpp

```diff
@@ -83,69 +87,56 @@ class DifferentialDriveModel {
-   *  requirements.
-   * \param state The state of the particle to which the motion will be applied.
-   * \param gen An uniform random bit generator object.
+   * \tparam ControlAction A tuple-like container matching the model's `control_action_type`.
+   * \param action Control action to condition the motion model with.
+   * \return a state sampling function satisfying \ref MotionSamplingPage.
    */
-  template <class Generator>
-  [[nodiscard]] state_type apply_motion(const state_type& state, Generator& gen) const {
-    static thread_local auto distribution = std::normal_distribution<double>{};
-    const auto first_rotation = Sophus::SO2d{distribution(gen, first_rotation_params_)};
-    const auto translation = Eigen::Vector2d{distribution(gen, translation_params_), 0.0};
-    const auto second_rotation = Sophus::SO2d{distribution(gen, second_rotation_params_)};
-    return state * Sophus::SE2d{first_rotation, Eigen::Vector2d{0.0, 0.0}} * Sophus::SE2d{second_rotation, translation};
-  }
-
-  /// \copydoc OdometryMotionModelInterface2d::update_motion(const Sophus::SE2d&)
-  void update_motion(const update_type& pose) {
-    if (last_pose_) {
-      const auto translation = pose.translation() - last_pose_.value().translation();
-      const double distance = translation.norm();
-      const double distance_variance = distance * distance;
-
-      const auto& previous_orientation = last_pose_.value().so2();
-      const auto& current_orientation = pose.so2();
-      const auto first_rotation =
-          distance > params_.distance_threshold
-              ? Sophus::SO2d{std::atan2(translation.y(), translation.x())} * previous_orientation.inverse()
-              : Sophus::SO2d{};
-      const auto second_rotation = current_orientation * previous_orientation.inverse() * first_rotation.inverse();
-
-      {
-        first_rotation_params_ = DistributionParam{
-            first_rotation.log(), std::sqrt(
-                                      params_.rotation_noise_from_rotation * rotation_variance(first_rotation) +
-                                      params_.rotation_noise_from_translation * distance_variance)};
-        translation_params_ = DistributionParam{
-            distance, std::sqrt(
-                          params_.translation_noise_from_translation * distance_variance +
-                          params_.translation_noise_from_rotation *
-                              (rotation_variance(first_rotation) + rotation_variance(second_rotation)))};
-        second_rotation_params_ = DistributionParam{
-            second_rotation.log(), std::sqrt(
-                                       params_.rotation_noise_from_rotation * rotation_variance(second_rotation) +
-                                       params_.rotation_noise_from_translation * distance_variance)};
-      }
-    }
-    last_pose_ = pose;
+  template <class ControlAction, typename = common_tuple_type_t<ControlAction, control_action_type>>
```

Comment:
This is the trick that allows tuple-like containers like the circular array to be used in place for standard tuples. This is arguably too complex. I convinced myself that was OK on the basis that (a) this is a core library abstraction, and (b) users don't need to go this far when implementing their models. They can fix all types and that will work too.

Reply:
> They can fix all types and that will work too.

+1